tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_persistence.c

```diff
@@ -0,0 +1,278 @@
+		while (!err) {
+			if (dentry.name[0] != 0) {
+				char innerpath[100] = { 0 };
+
+				strcpy(innerpath, dirpath);
+				strcat(innerpath, "/");
+				strcat(innerpath, dentry.name);
+
+				if (dentry.type == FS_DIR_ENTRY_DIR) {
+					err = read_dir(innerpath);
+				} else {
+					struct fs_file_t fp = { 0 };
+
+					LOG_INF("Size: %4d Entry: %s", dentry.size, innerpath);
+					err = fs_open(&fp, innerpath, FS_O_READ);
+					if (!err) {
+						char filedata[1000];
+						ssize_t n = fs_read(&fp, &filedata,
+								    ARRAY_SIZE(filedata));
+						LOG_HEXDUMP_INF(filedata, n, "raw data");
+
+						err = fs_close(&fp);
+					}
+				}
+				err = fs_readdir(&dirp, &dentry);
+				if (err) {
+					LOG_ERR("Fail to read directory %s", dirpath);
+					return fs_closedir(&dirp);
+				}
+			} else {
+				break;
+			}
+		}
+
+		return fs_closedir(&dirp);
+	}
+
+	return err;
+}
+
+struct key_checker_data {
+	const char *key;
+	const void *exp_value;
+	int exp_value_len;
+};
+
+
+typedef int (*key_value_checker_cb)(const char *key, size_t keylen,
+				    const uint8_t *data, ssize_t datalen);
+
+static bool key_checked;
+
+int set_load_direct_cb(const char *key,
+		       size_t len,
+		       settings_read_cb read_cb,
+		       void *cb_arg,
+		       void *param)
+{
+	uint8_t data[100];
+
+	LOG_INF("key: %s", key);
+
+	ssize_t bytes = read_cb(cb_arg, &data, sizeof(data));
+
+	LOG_HEXDUMP_INF(data, bytes, "data:");
+
+	struct key_checker_data *validate = (struct key_checker_data *)param;
+
+	if (strcmp(key, validate->key) == 0) {
+		key_checked = true;
+
+		if (!(bytes == validate->exp_value_len &&
+		      memcmp(data, validate->exp_value, bytes) == 0)) {
+			FAIL("Validation failed for key %s", key);
+			return -1;
+		}
+	}
+
+	return 0;
+}
+
+static int persistent_values_check(struct key_checker_data *validater)
+{
+
+	key_checked = false;
+	settings_load_subtree_direct("bt", set_load_direct_cb, validater);
+	if (!key_checked) {
+		FAIL("Not found. Key %s", validater->key);
+		return -1;
+	}
+
+	return 0;
+}
+
+/* these data structures are taken from various relavent modules to be checked */
+static struct net_key_val {
+	uint8_t kr_flag : 1,
+		kr_phase : 7;
+	uint8_t val[2][16];
+} __packed;
+
+static struct iv_val {
+	uint32_t iv_index;
+	uint8_t iv_update : 1,
+		iv_duration : 7;
+} __packed;
+
+static struct net_val {
+	uint16_t primary_addr;
+	uint8_t dev_key[16];
+} __packed;
+
+static void test_provisioning_sv_check_save(void)
+{
+	struct key_checker_data validater;
+
+	mount_settings_area();
+	bt_mesh_test_cfg_set(&save_cfg, WAIT_TIME);
+	bt_mesh_test_setup(false);
+
+	/* print raw dump of settings file */
+	char dir_path[100] = CONFIG_SETTINGS_FS_DIR;
+
+	read_dir(dir_path);
+
+	/** check stored provisioning data **/
+	/* check stored netkey matches with expected value */
+	char keypath[20];
+	struct net_key_val keyval = { 0 };
+
+	keyval.kr_flag = 0;
+	keyval.kr_phase = 0;
+	memcpy(keyval.val[0], save_cfg.net_key, sizeof(save_cfg.net_key));
```

Comment:
Shouldn't be both keys tested here?

Reply:
Initially I thought of using `mesh_test` functionality to implicitly test the loaded keys, but now added explicit check for all of provisioning data.